static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/python3
+
+import os
+import re
+import tempfile
+import yaml
+import contextlib
+from pathlib import Path
+
+from lib import util, results
+
+
+# Obtained from
+# https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#task
+ansible_reserved_keywords = [
+    'action', 'any_errors_fatal', 'args', 'async', 'become', 'become_exe', 'become_flags',
+    'become_method', 'become_user', 'changed_when', 'check_mode', 'collections', 'connection',
+    'debugger', 'delay', 'delegate_facts', 'delegate_to', 'diff', 'environment', 'failed_when',
+    'ignore_errors', 'ignore_unreachable', 'local_action', 'loop', 'loop_control',
+    'module_defaults', 'name', 'no_log', 'notify', 'poll', 'port', 'register', 'remote_user',
+    'retries', 'run_once', 'tags', 'throttle', 'timeout', 'until', 'vars', 'when', 'with_items'
+]
+
+
+@contextlib.contextmanager
+def select_all_rules(datastream):
+    """
+    Given an XML file path as 'datastream' enable all the rules in the datastream.
+
+    Use as a context manager, returning a filesystem path to the
+    modified datastream.
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules selected as {new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def traverse_blocks(task, allowed_modules):
+    block_set = set(['block'])
+    task_keywords = set(task.keys())
+
+    if len(task_keywords.intersection(block_set)) > 0:
+        for block_task in task['block']:
+            traverse_blocks(block_task, allowed_modules)
```

Comment:
This should probably be
```python
return traverse_blocks(block_task, allowed_modules)
```
otherwise you lose all info about nested blocks/tasks.

Reply:
Hmm, but it was your suggestion to remove it :D https://github.com/RHSecurityCompliance/contest/pull/22#discussion_r1235483926

Reply:
Okay, good point. This time around, I focused on recursion and didn't see the `for`. :smile: 

The last sentence is however correct. The overall logic should be something like
```python

found_allowed_modules = set()

for all items in the current task:
    if item is a block:
        found_allowed_modules += traverse_blocks(...)
    else:
        modules_in_current_item = ....
        if modules_in_current_item contains forbidden words:
            report fail
        else:
            found_allowed_modules += modules_in_current_item

return found_allowed_modules
```
Basically collect allowed modules from both nested blocks as well as current task list items.

(If you want to be fancy, you don't actually need `found_allowed_modules` at all, you can just do `yield from ...`, but I wanted to illustrate what I meant.)